src/Workspaces/Core/Portable/Workspace/IsolatedAnalyzerReferenceSet.Core.cs

```diff
@@ -0,0 +1,191 @@
+// See the LICENSE file in the project root for more information.
+
+#if NET
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Serialization;
+using System.Runtime.Loader;
+
+namespace Microsoft.CodeAnalysis;
+
+/// <summary>
+/// A set of <see cref="IsolatedAnalyzerReference"/>s and their associated shadow copy loader (which has its own <see
+/// cref="AssemblyLoadContext"/>).  As long as something is keeping this set alive, the ALC will be kept alive.  Once
+/// this set is dropped, the loader will be explicitly <see cref="IDisposable.Dispose"/>'d in its finalizer.
+/// </summary>
+internal sealed partial class IsolatedAnalyzerReferenceSet
+{
+    /// <summary>
+    /// Gate around <see cref="s_checksumToReferenceSet"/> to ensure it is only accessed and updated atomically.
+    /// </summary>
+    private static readonly SemaphoreSlim s_isolatedReferenceSetGate = new(initialCount: 1);
+
+    /// <summary>
+    /// Mapping from checksum for a particular set of assembly references, to the dedicated ALC and actual assembly
+    /// references corresponding to it.  As long as it is alive, we will try to reuse what is in memory.  But once it is
+    /// dropped from memory, we'll clean things up and produce a new one.
+    /// </summary>
+    private static readonly Dictionary<Checksum, WeakReference<IsolatedAnalyzerReferenceSet>> s_checksumToReferenceSet = [];
+
+    private static int s_sweepCount = 0;
+
+    /// <summary>
+    /// Final set of <see cref="AnalyzerReference"/> instances that will be passed through the workspace down to the compiler.
+    /// </summary>
+    public ImmutableArray<AnalyzerReference> AnalyzerReferences { get; }
+
+    /// <summary>
+    /// Dedicated loader with its own dedicated ALC that all analyzer references will load their <see
+    /// cref="System.Reflection.Assembly"/>s within.
+    /// </summary>
+    private readonly IAnalyzerAssemblyLoaderInternal _shadowCopyLoader;
+
+    private IsolatedAnalyzerReferenceSet(
+        ImmutableArray<AnalyzerReference> initialReferences,
+        IAnalyzerAssemblyLoaderProvider provider)
+    {
+        // Now make a fresh loader that uses that ALC that will ensure these references are properly isolated.
+        _shadowCopyLoader = provider.CreateNewShadowCopyLoader();
+
+        var builder = new FixedSizeArrayBuilder<AnalyzerReference>(initialReferences.Length);
+        foreach (var initialReference in initialReferences)
+        {
+            // If we already have an analyzer reference isolated to another ALC.  Fish out its underlying reference so
+            // we can rewrap it for the new ALC we're creating.  We don't want to continually wrap layers of isolated
+            // objects.
```

Comment:
this is a normal case.  consider just the following:

1. We add an AnalyzerReference for Foo.dll.  That makes an ANalyzerFileReference, which is wrapped with an IsolatedAnalyzerReference.
2. We then add an AnalyzerReference for Bar.dll.  This means we have to make an isolated set with both analyzer refernces.  however, the analyzerreference we have for Foo.dll is the IsoaltedAnalyzerRef made in step 1.  So we want to fish that out here to get back to the underlying file ref.

does that make sense?

Reply:
@jasonmalinowski  i just want to make sure we're good on this.

Reply:
I guess the surprising part to me is if we're adding a reference to Bar, that we're going _back_ to grab the stuff from the workspace instead of any sort of internal list that's owned by the ProjectSystemProject. Or just creating a fresh set of AnalzyerFileReferences and passing that to the isolation code. Put another way I imagine there'd be the "references that aren't isolated" and "the references that are isolated", and any list clearly is one or the other.

I'm not saying it's _bad_, but it seems like we're not being as strict with where references are coming from as I'd expect that we should be able to.

Reply:
i see.  let me try that out.